src/engine/engine.cpp

```diff
@@ -53,6 +69,47 @@ Expected<ThreadSafeModule> ExecEngine::optimize_module(
     return std::move(tsm);
 }
 
+void ExecEngine::Optimize(Module &llmod)
+{
+    LoopAnalysisManager LAM;
+    FunctionAnalysisManager FAM;
+    CGSCCAnalysisManager CGAM;
+    ModuleAnalysisManager MAM;
+
+    PassBuilder PB;
+    VectorizerParams::VectorizationFactor = 4;
+
+    PB.registerModuleAnalyses(MAM);
+    PB.registerCGSCCAnalyses(CGAM);
+    PB.registerFunctionAnalyses(FAM);
+    PB.registerLoopAnalyses(LAM);
+    PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
+
+    FunctionPassManager FPM;
+    FPM.addPass(SCCPPass());
+    FPM.addPass(ADCEPass());
+    FPM.addPass(
+        LoopSimplifyPass());  // needs to be followed by a simplifycfg pass
+    {
+        LoopPassManager LM;
+        LM.addPass(LoopRotatePass());
+        FPM.addPass(createFunctionToLoopPassAdaptor(std::move(LM)));
+    }
+    FPM.addPass(LoopUnrollPass());
+    FPM.addPass(SimplifyCFGPass());
+    FPM.addPass(ADCEPass());
```

Comment:
this is added twice

Reply:
is this intentional?

Reply:
Yes, I got this from someone's repo who recommended these passes in this order, including the duplicate, to help with vectorization. If you want it pared down lmk

Reply:
okay. then lets keep it

Reply:
can you tag the link where you got these passes in the code as a comment?